dds_web/security/project_user_keys.py

```diff
@@ -1,17 +1,168 @@
+
+
+def obtain_project_private_key(user, project):
+    project_key = models.ProjectUserKeys.query.filter_by(
+        project_id=project.id, user_id=user.username
+    ).first()
+    if project_key:
+        return decrypt_project_private_key_for_user(user, project_key.key)
+    raise KeyNotFoundError(project=project.public_id)
+
+
+def share_project_private_key_with_user(current_user, existing_user, project):
+    project_existing_user_key = models.ProjectUserKeys(
+        project_id=project.id,
+        user_id=existing_user.username,
+        key=encrypt_project_private_key_for_user(
+            existing_user, obtain_project_private_key(current_user, project)
+        ),
+    )
+    existing_user.project_user_keys.append(project_existing_user_key)
+    project.project_user_keys.append(project_existing_user_key)
+
+
+def generate_project_key_pair(user, project):
+    private_key = asymmetric.x25519.X25519PrivateKey.generate()
+    private_key_bytes = private_key.private_bytes(
+        encoding=serialization.Encoding.Raw,
+        format=serialization.PrivateFormat.Raw,
+        encryption_algorithm=serialization.NoEncryption(),
+    )
+    public_key_bytes = private_key.public_key().public_bytes(
+        encoding=serialization.Encoding.Raw,
+        format=serialization.PublicFormat.Raw,
+    )
+    project.public_key = public_key_bytes
+    for unit_user in user.unit.users:
+        encrypted_private_key = encrypt_project_private_key_for_user(unit_user, private_key_bytes)
+        project_user_key = models.ProjectUserKeys(
+            project_id=project.id, user_id=unit_user.username, key=encrypted_private_key
+        )
+        unit_user.project_user_keys.append(project_user_key)
+        project.project_user_keys.append(project_user_key)
+    del private_key_bytes
+    del private_key
+    gc.collect()
+
+
+def encrypt_with_aes(key, plaintext, aad=None):
+    """
+    Encrypts the plaintext with the AES algorithm using the key
+    :param key: symmetric key of the user
+    :param plaintext: a byte string
+    :param aad: Additional data that should be authenticated with the key, but is not encrypted. Can be None.
+    """
+    aesgcm = ciphers.aead.AESGCM(key)
+    nonce = os.urandom(12)
+    return {"nonce": nonce, "encrypted_key": aesgcm.encrypt(nonce, plaintext, aad)}
+
+
+def decrypt_with_aes(key, ciphertext, nonce, aad=None):
+    """
+    Decrypts the ciphertext with the AES algorithm using the key
+    :param key: symmetric key of the user
+    :param ciphertext: encrypted content
+    :param nonce: a unique value that has to be used only once to encrypt a given data with the given key
+    :param aad: Additional data that should be authenticated with the key, but is not encrypted. Can be None.
+    """
+    aesgcm = ciphers.aead.AESGCM(key)
+    return aesgcm.decrypt(nonce, ciphertext, aad)
+
+
+def encrypt_user_private_key(user, private_key):
+    user.temporary_key = ciphers.aead.AESGCM.generate_key(bit_length=256)
```

Comment:
Does this need a `db.session.commit` in order to be persisted?

Reply:
`encrypt_user_private_key` is called from `generate_user_key_pair` which currently is only called in `models.User.__init__`, and every time a user is created the `db.session.commit()` is called. So the commit isn't needed here at this point.